modules/mono/glue/GodotSharp/GodotSharp/Core/Projection.cs

```diff
@@ -0,0 +1,817 @@
+
+        public int GetPixelsPerMeter(int forPixelWidth)
+        {
+            Vector3 result = Xform(new Vector3(1, 0, -1));
+
+            return (int)((result.x * (real_t)0.5 + (real_t)0.5) * forPixelWidth);
+        }
+
+        public Plane GetProjectionPlane(Planes plane)
+        {
+            Plane newPlane = plane switch
+            {
+                Planes.Near => new Plane(x.w + x.z, y.w + y.z, z.z + z.y, w.w + w.z),
+                Planes.Far => new Plane(x.w - x.z, y.w - y.z, z.z - z.y, w.w - w.z),
+                Planes.Left => new Plane(x.w + x.x, y.w + y.x, z.w + z.x, w.w + w.x),
+                Planes.Top => new Plane(x.w - x.y, y.w - y.y, z.w - z.y, w.w - w.y),
+                Planes.Right => new Plane(x.w - x.x, y.w - y.x, z.w - z.x, w.w - w.x),
+                Planes.Bottom => new Plane(x.w + x.y, y.w + y.y, z.w + z.y, w.w + w.y),
+            };
+            newPlane.Normal = -newPlane.Normal;
+            return newPlane.Normalized();
+        }
+
+        public Vector2 GetViewportHalfExtends()
+        {
+            var res = GetProjectionPlane(Planes.Near).Intersect3(GetProjectionPlane(Planes.Right), GetProjectionPlane(Planes.Top));
+            return new Vector2(res.Value.x, res.Value.y);
+        }
+
+        public real_t GetZFar()
+        {
+            Plane newPlane = new Plane(x.w + x.z, y.w + y.z, z.z + z.y, w.w + w.z);
+            newPlane.Normal = -newPlane.Normal;
+            return newPlane.D;
+        }
+
+        public real_t GetZNear()
+        {
+            Plane newPlane = new Plane(x.w + x.z, y.w + y.z, z.w + z.z, -w.w - w.z);
+            return newPlane.Normalized().D;
+        }
+
+        public Projection Inverse()
+        {
+            Projection proj = this;
+            int i, j, k;
+            int[] pvt_i = new int[4];
+            int[] pvt_j = new int[4]; /* Locations of pivot matrix */
+            real_t pvt_val; /* Value of current pivot element */
+            real_t hold; /* Temporary storage */
+            real_t determinant = 1.0f;
+            for (k = 0; k < 4; k++)
+            {
+                /* Locate k'th pivot element */
+                pvt_val = proj[k][k]; /* Initialize for search */
+                pvt_i[k] = k;
+                pvt_j[k] = k;
+                for (i = k; i < 4; i++)
+                {
+                    for (j = k; j < 4; j++)
+                    {
+                        if (Mathf.Abs(proj[i][j]) > Mathf.Abs(pvt_val))
+                        {
+                            pvt_i[k] = i;
+                            pvt_j[k] = j;
+                            pvt_val = proj[i][j];
+                        }
+                    }
+                }
+
+                /* Product of pivots, gives determinant when finished */
+                determinant *= pvt_val;
+                if (Mathf.IsZeroApprox(determinant))
+                {
+                    GD.PushError("determinant is zero");
+                    return Zero;
```

Comment:
Since the C++ equivalent doesn't push an error here, I don't think C# should either.

https://github.com/godotengine/godot/blob/8cb51a642e7c34a73e7ed877ab46c041521205e5/core/math/projection.cpp#L619-L621